crates/telio-network-monitors/src/linux.rs

```diff
@@ -0,0 +1,66 @@
+use std::{io, mem};
+
+use libc::{bind, recvmsg, socket, AF_NETLINK, NETLINK_ROUTE, SOCK_RAW};
+use netlink::{Struct_nlmsghdr, Struct_sockaddr_nl};
+use telio_utils::{telio_log_debug, telio_log_trace, telio_log_warn};
+use tokio::task::JoinHandle;
+
+use crate::monitor::PATH_CHANGE_BROADCAST;
+
+/// Setup network path monitor for linux
+pub fn setup_monitor() {
+    let sock = open_netlink_socket().unwrap();
+    read_event(sock);
+}
+
+fn open_netlink_socket() -> Result<i32, i32> {
+    let sockfd = unsafe { socket(AF_NETLINK, SOCK_RAW, NETLINK_ROUTE) };
+    if sockfd < 0 {
+        return Err(sockfd);
+    }
+
+    let addr = netlink::Struct_sockaddr_nl {
+        nl_family: AF_NETLINK as u16,
+        nl_groups: netlink::RTM_NEWLINK | 0x1,
+        ..Default::default()
+    };
+
+    let res = unsafe {
+        bind(
+            sockfd,
+            &addr as *const _ as *const _,
+            mem::size_of::<netlink::Struct_sockaddr_nl>() as u32,
+        )
+    };
+
+    telio_log_debug!("Network monitor binding with socket {res}");
+    Ok(sockfd)
+}
+
+fn read_event(socket: i32) -> JoinHandle<io::Result<()>> {
+    tokio::spawn(async move {
+        let buf = vec![0; 1024 / mem::size_of::<netlink::Struct_nlmsghdr>()];
+        let mut sa = netlink::Struct_sockaddr_nl::default();
+        let mut iov = libc::iovec {
+            iov_base: buf.as_ptr() as *mut libc::c_void,
+            iov_len: buf.len(),
+        };
+
+        let mut msg = libc::msghdr {
+            msg_name: &mut sa as *mut _ as *mut _,
+            msg_namelen: mem::size_of_val(&sa) as u32,
+            msg_iov: &mut iov,
+            msg_iovlen: 1,
+            msg_control: std::ptr::null_mut(),
+            msg_controllen: 0,
+            msg_flags: 0,
+        };
+        loop {
+            let _ = unsafe { recvmsg(socket, &mut msg, 0) };
```

Comment:
Note that it is note super correct to have blocking IO inside of tokio task. But there is [spawn_blocking](https://docs.rs/tokio/latest/tokio/task/fn.spawn_blocking.html) to resolve this.